File: backend/app/utils/focal.py
def focal_from_detections(
    results: list[dict], width: int, height: int
) -> tuple[float, float] | None:
    """顔検出結果からフォーカルポイントを計算する。

    すべてのバウンディングボックスの和集合を使用し、
    複数の顔がある画像では1つだけでなく全検出顔の中心にフォーカルポイントを設定する。
    """
    if not results or width <= 0 or height <= 0:
        return None

    xmin = min(r["box"]["xmin"] for r in results)
    ymin = min(r["box"]["ymin"] for r in results)
    xmax = max(r["box"]["xmax"] for r in results)
    ymax = max(r["box"]["ymax"] for r in results)

    cx = (xmin + xmax) / 2
    # 頭部/髪のクロッピングを防ぐため顔ボックスの上部寄り (上から 1/3) にバイアス
    cy = ymin + (ymax - ymin) / 3

    focal_x = max(-1.0, min(1.0, (cx / width) * 2 - 1))
    focal_y = max(-1.0, min(1.0, 1 - (cy / height) * 2))
    return (focal_x, focal_y)

File: backend/rust-core/src/focal.rs
/// 顔検出のバウンディングボックス。座標系は元画像のピクセル座標。
pub struct BBox {
    pub xmin: f64,
    pub ymin: f64,
    pub xmax: f64,
    pub ymax: f64,
}

/// 1件の顔検出結果。
pub struct Detection {
    pub bbox: BBox,
}

/// `app.utils.focal.focal_from_detections` を移植したもの。
/// 顔検出結果からフォーカルポイントを計算する。
///
/// すべてのバウンディングボックスの和集合を使用し、
/// 複数の顔がある画像では1つだけでなく全検出顔の中心にフォーカルポイントを設定する。
pub fn focal_from_detections(
    detections: &[Detection],
    width: f64,
    height: f64,
) -> Option<(f64, f64)> {
    if detections.is_empty() || width <= 0.0 || height <= 0.0 {
        return None;
    }

    let mut xmin = f64::INFINITY;
    let mut ymin = f64::INFINITY;
    let mut xmax = f64::NEG_INFINITY;
    let mut ymax = f64::NEG_INFINITY;

    for d in detections {
        xmin = xmin.min(d.bbox.xmin);
        ymin = ymin.min(d.bbox.ymin);
        xmax = xmax.max(d.bbox.xmax);
        ymax = ymax.max(d.bbox.ymax);
    }

    let cx = (xmin + xmax) / 2.0;
    // 頭部/髪のクロッピングを防ぐため顔ボックスの上部寄り (上から 1/3) にバイアス
    let cy = ymin + (ymax - ymin) / 3.0;

    let focal_x = ((cx / width) * 2.0 - 1.0).clamp(-1.0, 1.0);
    let focal_y = (1.0 - (cy / height) * 2.0).clamp(-1.0, 1.0);

    Some((focal_x, focal_y))
}
